src/main/kotlin/com/sourcegraph/cody/context/ui/EnhancedContextPanel.kt

```diff
@@ -1,58 +1,257 @@
-          .addExtraAction(helpButton)
-          .setPreferredSize(Dimension(0, 30))
-          .setToolbarPosition(ActionToolbarPosition.LEFT)
-          .setScrollPaneBorder(BorderFactory.createEmptyBorder())
-          .setToolbarBorder(BorderFactory.createEmptyBorder())
-          .createPanel()
+
+    remoteContextNode.add(remoteRepoNode)
+    treeModel.reload()
+  }
+
+  @RequiresEdt
+  private fun expandAllNodes(rowCount: Int = tree.rowCount) {
+    for (i in 0 until tree.rowCount) {
+      tree.expandRow(i)
+    }
+
+    if (tree.getRowCount() != rowCount) {
+      expandAllNodes(tree.rowCount)
+    }
+  }
 
   init {
-    layout = VerticalFlowLayout(VerticalFlowLayout.BOTTOM, 0, 5, true, false)
-    tree.expandRow(0)
+    layout = VerticalFlowLayout(VerticalFlowLayout.BOTTOM, 0, 0, true, false)
+    tree.setModel(treeModel)
+    prepareTree()
+
+    val toolbarDecorator =
+        createDecorator(tree)
+            .disableUpDownActions()
+            .setVisibleRowCount(1)
+            .setScrollPaneBorder(BorderFactory.createEmptyBorder())
+            .setToolbarBorder(BorderFactory.createEmptyBorder())
+
+    if (!isDotComAccount()) {
+      toolbarDecorator.setAddActionName(
+          CodyBundle.getString("context-panel.button.add-remote-repo"))
+      toolbarDecorator.setAddAction {
+        AddRepositoryDialog(project) { repoUrl -> addRemoteRepository(repoUrl) }.show()
+        expandAllNodes()
+      }
+
+      toolbarDecorator.setRemoveActionName(
+          CodyBundle.getString("context-panel.button.remove-remote-repo"))
+      toolbarDecorator.setRemoveActionUpdater {
+        tree.selectionPath?.lastPathComponent is ContextTreeRemoteRepoNode
+      }
+      toolbarDecorator.setRemoveAction {
+        (tree.selectionPath?.lastPathComponent as? ContextTreeRemoteRepoNode)?.let { node ->
+          removeRemoteRepository(node)
+          expandAllNodes()
+        }
+      }
+    }
+
+    toolbarDecorator.addExtraAction(ReindexButton(project))
+
+    toolbarDecorator.addExtraAction(
+        ContextToolbarButton(
+            CodyBundle.getString("context-panel.button.save-default"),
+            AllIcons.Actions.SetDefault) {
+              HistoryService.getInstance(project).updateDefaultContextState(getContextState())
+            })
+
+    toolbarDecorator.addExtraAction(
+        ContextToolbarButton(
+            CodyBundle.getString("context-panel.button.restore-default"), AllIcons.General.Reset) {
+              HistoryService.getInstance(project)
+                  .updateContextState(chatSession.getInternalId(), contextState = null)
+              treeRoot.removeAllChildren()
+              prepareTree()
+            })
+
+    toolbarDecorator.addExtraAction(HelpButton())
+
+    val panel = toolbarDecorator.createPanel()
+
+    tree.addTreeExpansionListener(
+        object : TreeExpansionListener {
+          private fun resize() {
+            panel.preferredSize = Dimension(0, 30 + tree.rowCount * 20)
```

Comment:
`30 + tree.rowCount * 20` :magic_wand: :smile: 

Will this formula remain valid after we change the UI font size or font type? (Appearance > Use custom font)

Reply:
That is the part I mentioned on the standup when I said I will be happy if someone will came up with anything better :) We definitely need to improve it, it's just that it turned out to be much harder than it seems.
Since I have a moment now I will get back to it and check if I can fix it somehow.

Reply:
Thx!